File: dotfile/handlers/directory.py
"""Directory handler for creating directories."""

import os
from typing import Dict, Any

from ..core.handler import Status, CheckResult
from ..core.utils import expand_path


class DirectoryHandler:
    """Handler for creating directories."""
    
    def check(self, config: Dict[str, Any]) -> CheckResult:
        """Check if directory exists."""
        # Handle both single path and multiple paths
        paths = config.get("paths", [config.get("path")])
        if not isinstance(paths, list):
            paths = [paths]
        
        # Filter out None values
        paths = [p for p in paths if p]
        
        if not paths:
            return CheckResult(
                status=Status.ERROR,
                message="No paths specified"
            )
        
        # Check all paths
        missing = []
        errors = []
        
        for path_str in paths:
            path = expand_path(path_str)
            
            if path.exists():
                if not path.is_dir():
                    errors.append(f"{path} exists but is not a directory")
                # Check permissions if specified (only for single path)
                elif len(paths) == 1 and "mode" in config:
                    current_mode = oct(path.stat().st_mode)[-3:]
                    expected_mode = config["mode"].lstrip("0").zfill(3)
                    
                    if current_mode != expected_mode:
                        return CheckResult(
                            status=Status.MODIFIED,
                            message=f"Directory exists with wrong permissions: {path}"
                        )
            else:
                missing.append(str(path))
        
        if errors:
            return CheckResult(
                status=Status.ERROR,
                message="; ".join(errors)
            )
        elif missing:
            return CheckResult(
                status=Status.NOT_INSTALLED,
                message=f"Directories do not exist: {', '.join(missing)}"
            )
        else:
            return CheckResult(
                status=Status.INSTALLED,
                message="All directories exist"
            )
    
    def install(self, config: Dict[str, Any], dry_run: bool = False) -> bool:
        """Create directories."""
        # Handle both single path and multiple paths
        paths = config.get("paths", [config.get("path")])
        if not isinstance(paths, list):
            paths = [paths]
        
        # Filter out None values
        paths = [p for p in paths if p]
        
        if not paths:
            print("✗ No paths specified")
            return False
        
        if dry_run:
            print(f"Would create directories: {', '.join(paths)}")
            return True
        
        success = True
        permissions = config.get("permissions")
        parents = config.get("parents", True)
        
        for path_str in paths:
            path = expand_path(path_str)
            
            try:
                # Handle existing non-directory
                if path.exists() and not path.is_dir():
                    # Back up existing file
                    backup_num = 1
                    while True:
                        backup_path = path.parent / f"{path.name}.backup{backup_num}"
                        if not backup_path.exists():
                            break
                        backup_num += 1
                    
                    path.rename(backup_path)
                    print(f"  → Backed up existing file to {backup_path}")
                
                # Create directory
                path.mkdir(parents=parents, exist_ok=True)
                
                # Set permissions if specified
                if permissions:
                    try:
                        # Handle both "755" and "0755" formats
                        mode_str = permissions.lstrip("0")
                        mode_int = int(mode_str, 8)
                        os.chmod(path, mode_int)
                    except ValueError:
                        print(f"  ⚠️  Invalid permissions: {permissions}")
                
                print(f"✓ Created directory: {path}")
                
            except Exception as e:
                print(f"✗ Failed to create directory {path}: {e}")
                success = False
        
        return success
    
    def uninstall(self, config: Dict[str, Any], dry_run: bool = False) -> bool:
        """Remove directories if empty."""
        # Handle both single path and multiple paths
        paths = config.get("paths", [config.get("path")])
        if not isinstance(paths, list):
            paths = [paths]
        
        # Filter out None values
        paths = [p for p in paths if p]
        
        if dry_run:
            print(f"Would remove directories (if empty): {', '.join(paths)}")
            return True
        
        success = True
        
        # Process in reverse order (deepest first)
        for path_str in reversed(paths):
            path = expand_path(path_str)
            
            try:
                if path.exists() and path.is_dir():
                    # Only remove if empty
                    try:
                        path.rmdir()
                        print(f"✓ Removed empty directory: {path}")
                    except OSError:
                        print(f"⚠️  Directory not empty, keeping: {path}")
                else:
                    print(f"✓ Directory doesn't exist: {path}")
                
            except Exception as e:
                print(f"✗ Error removing directory {path}: {e}")
                success = False
        
        return success

File: dots/source/handlers/directory.ts
import * as fs from "fs";
import type { CheckResult, DirectoryConfig, Handler, MacOSConfig } from "../lib/types.js";
import { expandPath, appendLog, backupFile } from "../lib/shell.js";
import { isDirectory } from "../lib/types.js";

export const directoryHandler: Handler = {
	async check(config: MacOSConfig, name: string): Promise<CheckResult> {
		if (!isDirectory(config)) {
			return { status: "error", message: "Invalid directory config" };
		}

		const dirConfig = config as DirectoryConfig;
		const paths = getPaths(dirConfig);

		if (paths.length === 0) {
			return { status: "error", message: "No paths specified" };
		}

		const missingPaths: string[] = [];
		const wrongPermsPaths: string[] = [];
		const mode = dirConfig.mode || dirConfig.permissions;

		for (const p of paths) {
			const expandedPath = expandPath(p);

			if (!fs.existsSync(expandedPath)) {
				missingPaths.push(p);
				continue;
			}

			const stats = fs.statSync(expandedPath);
			if (!stats.isDirectory()) {
				return {
					status: "modified",
					message: `${p} exists but is not a directory`,
				};
			}

			// Check permissions if specified (only for single path)
			if (mode && paths.length === 1) {
				const currentMode = (stats.mode & 0o777).toString(8);
				const expectedMode = mode.replace(/^0/, "");
				if (currentMode !== expectedMode) {
					wrongPermsPaths.push(p);
				}
			}
		}

		if (missingPaths.length > 0) {
			return {
				status: "not_installed",
				message: `Missing: ${missingPaths.join(", ")}`,
			};
		}

		if (wrongPermsPaths.length > 0) {
			return {
				status: "modified",
				message: `Wrong permissions: ${wrongPermsPaths.join(", ")}`,
			};
		}

		return { status: "installed", message: "All directories exist" };
	},

	async install(
		config: MacOSConfig,
		name: string,
		dryRun = false,
	): Promise<boolean> {
		if (!isDirectory(config)) {
			return false;
		}

		const dirConfig = config as DirectoryConfig;
		const paths = getPaths(dirConfig);
		const mode = dirConfig.mode || dirConfig.permissions;
		const parents = dirConfig.parents !== false; // Default to true

		if (paths.length === 0) {
			await appendLog(`No paths specified for ${name}`);
			return false;
		}

		for (const p of paths) {
			const expandedPath = expandPath(p);

			if (dryRun) {
				await appendLog(`[DRY-RUN] Would create directory: ${expandedPath}`);
				continue;
			}

			try {
				// Check if a file exists at this path
				if (fs.existsSync(expandedPath)) {
					const stats = fs.statSync(expandedPath);
					if (!stats.isDirectory()) {
						// Backup the file
						const backupPath = await backupFile(expandedPath);
						if (backupPath) {
							await appendLog(`Backed up file to: ${backupPath}`);
						}
					} else {
						// Directory already exists, just set permissions if needed
						if (mode) {
							const modeNum = parseInt(mode.replace(/^0/, ""), 8);
							fs.chmodSync(expandedPath, modeNum);
							await appendLog(`Set permissions on ${expandedPath}: ${mode}`);
						}
						continue;
					}
				}

				// Create directory
				fs.mkdirSync(expandedPath, {
					recursive: parents,
					mode: mode ? parseInt(mode.replace(/^0/, ""), 8) : 0o755,
				});
				await appendLog(`Created directory: ${expandedPath}`);
			} catch (error) {
				await appendLog(`Failed to create directory ${expandedPath}: ${error}`);
				return false;
			}
		}

		return true;
	},

	async uninstall(
		config: MacOSConfig,
		name: string,
		dryRun = false,
	): Promise<boolean> {
		if (!isDirectory(config)) {
			return false;
		}

		const dirConfig = config as DirectoryConfig;
		const paths = getPaths(dirConfig);

		// Process in reverse order (deepest first)
		const reversedPaths = [...paths].reverse();

		for (const p of reversedPaths) {
			const expandedPath = expandPath(p);

			if (!fs.existsSync(expandedPath)) {
				continue;
			}

			try {
				const entries = fs.readdirSync(expandedPath);
				if (entries.length > 0) {
					await appendLog(`Skipping non-empty directory: ${expandedPath}`);
					continue;
				}

				if (dryRun) {
					await appendLog(`[DRY-RUN] Would remove empty directory: ${expandedPath}`);
					continue;
				}

				fs.rmdirSync(expandedPath);
				await appendLog(`Removed empty directory: ${expandedPath}`);
			} catch (error) {
				await appendLog(`Failed to remove directory ${expandedPath}: ${error}`);
				// Don't fail - directories may contain user data
			}
		}

		return true;
	},
};

/**
 * Get paths from directory config
 */
function getPaths(config: DirectoryConfig): string[] {
	if (config.paths) {
		return Array.isArray(config.paths) ? config.paths : [config.paths];
	}
	if (config.path) {
		return [config.path];
	}
	return [];
}
